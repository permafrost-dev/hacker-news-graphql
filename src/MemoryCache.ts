export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache {
    public map: CacheMap = {};

    public has(key: string): boolean {
        this.purge();

        return typeof this.map[key] !== 'undefined';
    }

    public get(key: string): any {
        this.purge();

        return this.map[key]?.value ?? null;
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[key] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }
}
