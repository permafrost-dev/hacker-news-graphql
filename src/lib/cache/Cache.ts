/* eslint-disable no-unused-vars */

export abstract class Cache {
    public prefix = '';

    public destroy() {
        //
    }
    public abstract has(key: string): Promise<boolean>;

    public abstract get(key: string): Promise<any | null>;

    public abstract put(key: string, value: any, ttlSeconds: number): void;

    public abstract purge();

    public abstract clear(): boolean;
}
