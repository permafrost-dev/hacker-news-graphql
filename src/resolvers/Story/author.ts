import axios from 'axios';

export function getResolver(cache: any): any {
    const authorResolver = async parent => {
        const userCacheKey = `user:${parent.by}`;

        const hasKey = await cache.has(userCacheKey);
        if (hasKey) {
            return cache.get(userCacheKey);
        }

        const url = `${process.env.HACKERNEWS_API_URL}/user/${parent.by}.json`;
        const resp = await axios.get(url);

        cache.put(userCacheKey, resp.data, 3600);

        return new Promise(resolve => resolve(cache.get(userCacheKey)));
    };

    return authorResolver;
}
