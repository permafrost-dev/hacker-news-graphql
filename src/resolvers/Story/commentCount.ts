/* eslint-disable no-unused-vars */

export function getResolver(cache: any): any {
    const commentCountResolver = parent => {
        return parent.descendants || 0;
    };

    return commentCountResolver;
}
