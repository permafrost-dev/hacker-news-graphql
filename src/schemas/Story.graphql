type Story {
    author: User
    by: String
    commentCount: Int!
    comments(first: Int!): [Comment]
    descendants: Int
    id: Int!
    kids: [Int]
    score: Int
    time: Date
    title: String
    url: String
    type: String
}
