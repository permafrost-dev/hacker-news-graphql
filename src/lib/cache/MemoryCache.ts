import { Cache } from './Cache';

export interface CacheItem {
    ts: number;
    expires: number;
    value: any;
}

export interface CacheMap {
    [key: string]: CacheItem;
}

export class MemoryCache extends Cache {
    public map: CacheMap = {};

    constructor(keyPrefix = '') {
        super();

        if (keyPrefix.length) {
            keyPrefix += ':';
        }

        this.prefix = keyPrefix;
    }

    protected makeKey(key: string): string {
        return `${this.prefix}${key}`;
    }

    public async has(key: string): Promise<boolean> {
        this.purge();

        return new Promise(resolve => resolve(typeof this.map[this.makeKey(key)] !== 'undefined'));
    }

    public async get(key: string): Promise<any | null> {
        this.purge();

        const result = this.map[this.makeKey(key)]?.value ?? null;

        return new Promise(resolve => resolve(result));
    }

    public put(key: string, value: any, ttlSeconds: number) {
        this.map[this.makeKey(key)] = {
            ts: new Date().getTime(),
            expires: new Date().getTime() + ttlSeconds * 1000,
            value: value,
        };
    }

    public purge() {
        for (const prop in this.map) {
            if (new Date().getTime() >= this.map[prop].expires) {
                delete this.map[prop];
                continue;
            }
        }
    }

    public clear() {
        //
    }
}
