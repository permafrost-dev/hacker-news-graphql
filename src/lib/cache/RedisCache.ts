import { Cache } from './Cache';
const Redis = require('ioredis');

export class RedisCache extends Cache {
    public redis: typeof Redis;

    constructor(keyPrefix = '') {
        super();

        if (keyPrefix.length) {
            keyPrefix += ':';
        }

        this.prefix = keyPrefix;

        this.redis = new Redis(); //{ keyPrefix: keyPrefix });
    }

    public destroy() {
        this.redis.disconnect();
    }

    public async has(key: string): Promise<boolean> {
        const result = await this.redis.exists(key);

        return new Promise(resolve => resolve(result === 1));
    }

    public async get(key: string): Promise<any | null> {
        const result = await this.redis.get(key);

        return new Promise(resolve => resolve(JSON.parse(result)));
    }

    public put(key: string, value: any, ttlSeconds: number): void {
        this.redis.set(key, JSON.stringify(value), 'EX', ttlSeconds);
    }

    public purge() {
        //
    }

    public clear() {
        //
    }
}
