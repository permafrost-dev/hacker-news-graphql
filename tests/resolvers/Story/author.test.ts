/* eslint-disable no-undef */

import { MemoryCache } from '../../../src/lib/cache/MemoryCache';
import { getResolver } from '../../../src/resolvers/Story/author';

let cache: MemoryCache;
let resolver: any;

process.env.HACKERNEWS_API_URL = 'http://127.0.0.1:3031';

beforeEach(() => {
    cache = new MemoryCache();
    resolver = getResolver(cache);
});

it('it retrieves the user data and returns it', async () => {
    const parent = { by: 'test' };
    const userData = await resolver(parent);

    expect(userData).toMatchSnapshot();
});

it('it retrieves the user data and caches it', async () => {
    await resolver({ by: 'test' });

    expect(cache.has('user:test')).toBeTruthy();
});
