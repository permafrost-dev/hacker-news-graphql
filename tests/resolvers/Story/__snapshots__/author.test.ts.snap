// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it retrieves the user data and returns it 1`] = `
Object {
  "id": "test",
  "karma": 1234,
}
`;
